src/core/backbone_router/multicast_listeners_table.hpp

```diff
@@ -0,0 +1,226 @@
+ */
+
+#ifndef MULTICAST_LISTENERS_TABLE_HPP
+#define MULTICAST_LISTENERS_TABLE_HPP
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+
+#include <openthread/backbone_router_ftd.h>
+
+#include "common/notifier.hpp"
+#include "common/time.hpp"
+#include "net/ip6_address.hpp"
+
+namespace ot {
+
+namespace BackboneRouter {
+
+/**
+ * This class implements the definitions for Multicast Listeners Table.
+ *
+ */
+class MulticastListenersTable : public InstanceLocator
+{
+    class IteratorBuilder;
+
+public:
+    /**
+     * This class represents a Multicast Listener entry.
+     *
+     */
+    class Listener : public Clearable<Listener>
+    {
+        friend class MulticastListenersTable;
+
+    public:
+        /**
+         * This constructor initializes the `Listener` object.
+         *
+         */
+        Listener(void) { Clear(); }
+
+        /**
+         * This method returns the listened Ip6 multicast address.
+         *
+         * @returns The listened Ip6 multicast address.
+         *
+         */
+        const Ip6::Address &GetAddress(void) const { return mAddress; }
+
+        /**
+         * This method returns the expire time of the listened Ip6 multicast address.
+         *
+         * @returns The expire time.
+         *
+         */
+        const TimeMilli GetExpireTime(void) const { return mExpireTime; }
+
+    private:
+        void SetAddress(const Ip6::Address &aAddress) { mAddress = aAddress; }
+        void SetExpireTime(TimeMilli aExpireTime) { mExpireTime = aExpireTime; }
+
+        static bool IsAfter(const Listener &aOne, const Listener &aOther)
+        {
+            return aOne.GetExpireTime() > aOther.GetExpireTime();
+        }
+
+        Ip6::Address mAddress;
+        TimeMilli    mExpireTime;
+    };
+
+    /**
+     * This constructor initializes the Multicast Listeners Table.
+     *
+     * @param[in] aInstance  A reference to the OpenThread instance.
+     *
+     */
+    explicit MulticastListenersTable(Instance &aInstance)
+        : InstanceLocator(aInstance)
+        , mNumValidListeners(0)
+    {
+    }
+
+    /**
+     * This method adds an Ip6 multicast address with a given expire time.
+     *
+     * @param[in] aAddress     The Ip6 multicast address.
+     * @param[in] aExpireTime  The expire time.
+     *
+     * @retval OT_ERROR_NONE          If the Ip6 multicast address was successfully added.
+     * @retval OT_ERROR_INVALID_ARGS  If the Ip6 multicast address was invalid.
+     * @retval OT_ERROR_NO_BUFS       No space available to save the Ip6 multicast address.
+     *
+     */
+    otError Add(const Ip6::Address &aAddress, TimeMilli aExpireTime);
+
+    /**
+     * This method removes a given Ip6 multicast address.
+     *
+     * @param[in] aAddress  The Ip6 multicast address to remove.
+     *
+     */
+    void Remove(const Ip6::Address &aAddress);
+
+    /**
+     * This method removes expired Ip6 multicast addresses.
```

Comment:
```suggestion
     * This method removes the expired Ip6 multicast addresses.
```

Reply:
There could be multiple expired addresses. No need to use `the` ?

Reply:
yes